advanced_server.rb

```diff
@@ -0,0 +1,312 @@
+require 'sinatra'
+require 'octokit'
+require 'dotenv/load'
+require 'git'
+require 'json'
+require 'openssl' # Used to verify the webhook signature
+require 'jwt'     # Used to authenticate a GitHub App
+require 'time'    # Used to get ISO 8601 representation of a Time object
+require 'logger'
+
+set :port, 3000
+set :bind, '0.0.0.0'
+
+
+# This is template code to create a GitHub App server.
+# You can read more about GitHub Apps here: # https://developer.github.com/apps/
+#
+# On its own, this app does absolutely nothing, except that it can be installed.
+# It's up to you to add fun functionality!
+# You can check out one example in advanced_server.rb.
+#
+# This code is a Sinatra app, for two reasons:
+#   1. Because the app will require a landing page for installation.
+#   2. To easily handle webhook events.
+#
+#
+# Of course, not all apps need to receive and process events!
+# Feel free to rip out the event handling code if you don't need it.
+#
+# Have fun!
+#
+
+class GHAapp < Sinatra::Application
+
+  # Expects that the private key has been set as an environment variable in
+  # PEM format using the following command to replace newlines with the
+  # literal `\n`:
+  #   export GITHUB_PRIVATE_KEY=`awk '{printf "%s\\n", $0}' private-key.pem`
+  #
+  # Converts the newlines
+  PRIVATE_KEY = OpenSSL::PKey::RSA.new(ENV['GITHUB_PRIVATE_KEY'].gsub('\n', "\n"))
+
+  # Your registered app must have a secret set. The secret is used to verify
+  # that webhooks are sent by GitHub.
+  WEBHOOK_SECRET = ENV['GITHUB_WEBHOOK_SECRET']
+
+  # The GitHub App's identifier (type integer) set when registering an app.
+  APP_IDENTIFIER = ENV['GITHUB_APP_IDENTIFIER']
+
+  # Turn on Sinatra's verbose logging during development
+  configure :development do
+    set :logging, Logger::DEBUG
+  end
+
+
+  # Before each request to the `/event_handler` route
+  before '/event_handler' do
+    get_payload_request(request)
+    verify_webhook_signature
+    authenticate_app
```

Comment:
This will `404` for Check suite `created` webhooks for another app. The `if @payload['check_run']['app']['id'].to_s === APP_IDENTIFIER` check should get moved here.

Reply:
@osowskit ah I see the problem there. Should we move the `authenticate_app` and `authenticate_installation` methods into each `when` statement below? We'll have to duplicate the code for each `when` statement, but new tokens would only be created when they had passed all of the checks.

Reply:
Circling back to this comment, I realized that this update I added before the `authenticate_app` method wouldn't work:

```
halt 400 unless @payload[request.env['HTTP_X_GITHUB_EVENT']]['app']['id'].to_s === APP_IDENTIFIER
```

The APP_ID is inside of the event object, so it didn't seem very clean to check for more than one expected event. I think the app will `401` when the payload is for another app in the `verify_webhook_signature` before the `authenticate_app` method is called. I'm going to test how this works in the AM when I have two apps installed on the same repository creating check runs. If you see a good workaround, I'm all ears.



```
{
  "action": "completed",
  "check_suite": {
        "app": {
                    "id": 17104
         }
    }
}
```